juju/juju
# Adding update-series command by application.

## Description of change

The update-series command will allow a user to juju of a new series for an application or a machine.  This initial commit covers the application case where updating the series of an application allows for new units of the application to use the new series.

The new series must be in the supported series list of the related charm.  As well, charms subordinate
to the one specified must also support the new series.  A --force flag allows this behavior to be over ridden.

## QA steps

1. Deploy two charms, one a subordinate.  It's recommended to have 1 series supported by the principal charm not supported by the subordinate.
2. juju update-series <application> <series>; where the series is supported by both the principal and subordinates.  This should succeed.
3. juju add-unit <application>; the new unit should be provisioned with the series used in step 2.
4. juju update-series <application> <series>; where the series is supported by the principal and not the subordinate.  This should fail.
5. juju update-series <application> <series> --force; where the series is supported by the principal and not the subordinate.  This should succeed.
6. juju update-series <application> <series>; where the series is not supported by the principal.  This should fail.
7. juju update-series <application> <series> --force; where the series is not supported by the principal.  This should succeed.

At any time, the series to be used by an application can be found in the output of "juju status <application> --format yaml"

## Documentation changes

Yes, the online list of commands needs to be updated.

## Bug reference

This is a planned project, however a feature request bug was filed as well:
https://bugs.launchpad.net/juju/+bug/1704135
